### PR TITLE
Vectorise string.Replace(char, char)

### PR DESCRIPTION
**Pull request by Gnbrkm41**
* Replace bits of existing unsafe code to IndexOf, which is already vectorised
* Vectorise the replacement bit

I will PR more tests/benchmarks that test both the non-vectorised and vectorised paths in corefx/performance repo. (In the meantime, I have verified that the tests pass locally.)

Also, I am aware that the closure of the repo for consolidation is due; I am happy to open the PR again once the repo is open, if this cannot be merged before the closure.

Tested with four cases (no characters replaced / some characters replaced, with two different strings), and the results seem to suggest slight slowdown for strings that are too small to be vectorised & has something to be replaced (here it appears to be around 20% slower), but for longer strings it seems to be better:
```
summary:
better: 2, geomean: 1.293
worse: 1, geomean: 1.193
total diff: 3
```

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Tests.Perf_String.Replace_Char(text: "Hello", oldChar: 'l', newChar: '!') |      1.19 |             9.44 |            11.27 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Tests.Perf_String.Replace_Char(text: "This is a very nice sentence", oldChar: 'i', newChar: 'I') |      1.33 |             9.36 |             7.05 |         |
| System.Tests.Perf_String.Replace_Char(text: "This is a very nice sentence", oldChar: 'z', newChar: 'y') |      1.26 |            22.62 |            17.95 | several?|

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Why not use the normal `Vector<T>` APIs for this? As far as I can tell we're not relying on specific x86 intrinsic instructions. Using `Vector<T>` also allows this code to "just work" on ARM platforms.

**Comment by Gnbrkm41** [comment]
New perf numbers after running it on an AVX2 supported machine:
```
C:\Users\gotos\source\repos\performance\src\tools\ResultsComparer>dotnet run --base C:\Users\gotos\desktop\replace-before --diff C:\users\gotos\desktop\replace-after --threshold 5%
summary:
better: 4, geomean: 2.626
worse: 1, geomean: 1.204
total diff: 5
```

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| `"Hello".Replace('l', '!')` |      1.20 |            10.49 |            12.63 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| `"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut id tempor metus.".Replace('z', 'y')` |      5.19 |            46.50 |             8.97 |         |
| `"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut id tempor metus.".Replace('e', 'E')` |      2.79 |            66.68 |            23.93 |         |
| `"This is a very nice sentence".Replace('z', 'y')` |      2.21 |            14.65 |             6.62 |         |
|  `"This is a very nice sentence".Replace('i', 'I')` |      1.49 |            28.97 |            19.47 | several?|

(Also ran `"Hello".Replace('a', 'b')`, but there was no difference at 5% threshold)

File: src/System.Private.CoreLib/shared/System/String.Manipulation.cs
Comment (GrabYourPitchforks) [inline src/System.Private.CoreLib/shared/System/String.Manipulation.cs]: Can perform the comparison directly without needing a subtraction operation.

```suggestion
                while (remainingLength >= Vector<ushort>.Count);
```

File: src/System.Private.CoreLib/shared/System/String.Manipulation.cs
Comment (GrabYourPitchforks) [inline src/System.Private.CoreLib/shared/System/String.Manipulation.cs]: This may end up reading from `pSrc` twice. Would be more efficient to read from `pSrc` into a local once, perform the comparison, then store one of the two locals (`newChar` or `yourNewTempLocal`) into `pDst`.

**Review by GrabYourPitchforks — CHANGES_REQUESTED** [review]

File: src/System.Private.CoreLib/shared/System/String.Manipulation.cs
Comment (Gnbrkm41) [reply src/System.Private.CoreLib/shared/System/String.Manipulation.cs]: It does seem to read from `pSrc` twice indeed.

**Review by GrabYourPitchforks — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
Hi @Gnbrkm41 - thanks for doing this work! Just wanted to give you a heads up that we're undergoing a repo consolidation at the moment. There's no immediate action required on your part here. We're waiting for the dust to settle a bit before we migrate + merge these PRs.

**Comment by Gnbrkm41** [comment]
Since dotnet/runtime repo is open, I will migrate the PR over to there and close this PR.

**Pull request closed without merging**
